File: sdks/python/template_test.py
import datetime

from cyanprintsdk.domain.core.cyan import Cyan, CyanProcessor, CyanGlob, GlobType
from cyanprintsdk.domain.core.deterministic import IDeterminism
from cyanprintsdk.domain.core.inquirer import IInquirer
from cyanprintsdk.main import start_template_with_fn
from cyanprintsdk.domain.core.question import (
    TextQ,
    CheckboxQ,
    ConfirmQ,
    DateQ,
    PasswordQ,
    SelectQ,
)


async def template(i: IInquirer, d: IDeterminism) -> Cyan:
    name = await i.text("What is your name?", "q1")

    age = await i.textQ(
        TextQ(
            message="What is your age?",
            id="q2",
            desc="The age of the person",
            default="20",
            initial="18",
            validate=lambda x: "" if x.isdigit() else "Needs to be a number",
        )
    )

    food = await i.checkbox(
        "What is your food?", "q3", ["apple", "orange", "pear", "honeydew"]
    )

    color = await i.checkboxQ(
        CheckboxQ(
            message="What is your color?",
            id="q4",
            options=["red", "blue", "green", "yellow"],
            desc="The color of the person",
        )
    )

    beef_ok = await i.confirm("Is beef ok?", "q5")

    pork_ok = await i.confirmQ(
        ConfirmQ(
            message="Is pork ok?",
            id="q6",
            desc="Whether the person likes pork",
            default=True,
            error_message="You must tell me if you like pork",
        )
    )

    birthday = await i.date_select("What is your birthday", "q7", "State your birthday")

    def validate_december_date(x):
        # If it's already a date object
        if isinstance(x, datetime.date):
            return "" if x.month == 12 else "Needs to be in December"

        # If it's a string, try to parse it
        if isinstance(x, str):
            try:
                date_obj = datetime.datetime.strptime(x, "%Y-%m-%d").date()
                return "" if date_obj.month == 12 else "Needs to be in December"
            except ValueError:
                return "Needs to be a valid date format (YYYY-MM-DD)"

        # Otherwise, it's an invalid type
        return "Needs to be a date"

    mum_birthday = await i.date_selectQ(
        DateQ(
            message="What is your mum's birthday?",
            id="q8",
            desc="The birthday of your mum",
            default=datetime.date(1980, 1, 1),
            max_date=datetime.date(2025, 12, 31),
            min_date=datetime.date(1900, 1, 1),
            validate=validate_december_date,
        )
    )

    password = await i.password("What is your password?", "q9")

    pin = await i.passwordQ(
        PasswordQ(
            message="What is your pin?",
            id="q10",
            desc="The pin of the person",
            confirmation=True,
            validate=lambda x: (
                "" if any(c.isupper() for c in x) else "You must have a capital letter"
            ),
        )
    )

    car = await i.select(
        "What is your favourite car?",
        "q11",
        [
            "BMW",
            "Mercedes",
            "Audi",
            "Volkswagen",
            "Toyota",
            "Honda",
            "Ford",
            "Chevrolet",
            "Nissan",
            "Hyundai",
            "Kia",
            "Volvo",
            "Jeep",
            "Land Rover",
            "Lexus",
            "Mazda",
            "Mercury",
            "Mitsubishi",
            "Pontiac",
            "Saab",
            "Skoda",
            "Suzuki",
            "Toyota",
            "Volkswagen",
            "Volvo",
        ],
        "State your favourite car",
    )

    plane = await i.selectQ(
        SelectQ(
            message="What is your favourite plane?",
            id="q12",
            desc="The plane of the person",
            options=[
                "Boeing",
                "Airbus",
                "Bombardier",
                "Embraer",
                "Cessna",
                "Piper",
                "Hawker",
                "Dassault",
                "Lockheed",
                "Beechcraft",
                "Gulfstream",
                "Cessna",
                "Piper",
                "Hawker",
                "Dassault",
                "Lockheed",
                "Beechcraft",
                "Gulfstream",
            ],
        )
    )

    t = d.get("time", lambda: "7")

    c = Cyan(
        plugins=[],
        processors=[
            CyanProcessor(
                name="hello",
                files=[
                    CyanGlob(
                        root=None,
                        glob="**/*.*",
                        exclude=[],
                        type=GlobType.Template,
                    )
                ],
                config={
                    "Name": name,
                    "Age": age,
                    "Food": food,
                    "Color": color,
                    "Beef": beef_ok,
                    "Pork": pork_ok,
                    "Birthday": birthday,
                    "MumBirthday": mum_birthday,
                    "Password": password,
                    "Pin": pin,
                    "Car": car,
                    "Plane": plane,
                    "Time": t,
                },
            )
        ],
    )
    return c


start_template_with_fn(template)

File: sdks/node/template_test.ts
import { QuestionType, StartTemplateWithLambda, GlobType, type Cyan } from './src/main.js';

StartTemplateWithLambda(async (i, d) => {
  const name = await i.text('What is your name?', 'q1');

  const age = await i.text({
    id: 'q2',
    message: 'What is your age?',
    type: QuestionType.Text,
    desc: 'The age of the person',
    default: '20',
    initial: '18',
    validate: x => {
      try {
        const num = Number(x);
        if (Number.isNaN(num)) return 'Needs to be a number';
        return null;
      } catch (e) {
        return 'Needs to be a number';
      }
    },
  });

  const food = await i.checkbox('What is your food?', ['apple', 'orange', 'pear', 'honeydew'], 'q3');

  const color = await i.checkbox({
    id: 'q4',
    message: 'What is your color?',
    options: ['red', 'blue', 'green', 'yellow'],
    desc: 'The color of the person',
    type: QuestionType.Checkbox,
  });

  const beefOk = await i.confirm('Is beef ok?', 'q5');

  const porkOk = await i.confirm({
    id: 'q6',
    message: 'Is pork ok?',
    desc: 'Whether the person likes pork',
    type: QuestionType.Confirm,
    default: true,
    errorMessage: 'You must tell me if you like pork',
  });

  const birthday = await i.dateSelect('What is your birthday', 'q7', 'State your birthday');

  const mumBirthday = await i.dateSelect({
    id: 'q8',
    message: "What is your mum's birthday?",
    desc: 'The birthday of your mum',
    type: QuestionType.DateSelect,
    default: new Date('1980-01-01'),
    maxDate: new Date('2025-12-31'),
    minDate: new Date('1900-01-01'),
    validate: x => {
      try {
        const date = new Date(x);
        if (Number.isNaN(date.getTime())) return 'Needs to be a date';
        return date.getMonth() === 11 ? null : 'Needs to be in December';
      } catch (e) {
        return 'Needs to be a date';
      }
    },
  });

  const password = await i.password('What is your password?', 'q9');

  const pin = await i.password({
    id: 'q10',
    message: 'What is your pin?',
    desc: 'The pin of the person',
    confirmation: true,
    type: QuestionType.Password,
    validate: x => (/[A-Z]/.test(x) ? null : 'You must have a capital letter'),
  });

  const car = await i.select(
    'What is your favourite car?',
    [
      'BMW',
      'Mercedes',
      'Audi',
      'Volkswagen',
      'Toyota',
      'Honda',
      'Ford',
      'Chevrolet',
      'Nissan',
      'Hyundai',
      'Kia',
      'Volvo',
      'Jeep',
      'Land Rover',
      'Lexus',
      'Mazda',
      'Mercury',
      'Mitsubishi',
      'Pontiac',
      'Saab',
      'Skoda',
      'Suzuki',
      'Toyota',
      'Volkswagen',
      'Volvo',
    ],
    'q11',
    'State your favourite car',
  );

  const plane = await i.select({
    id: 'q12',
    message: 'What is your favourite plane?',
    desc: 'The plane of the person',
    type: QuestionType.Select,
    options: [
      'Boeing',
      'Airbus',
      'Bombardier',
      'Embraer',
      'Cessna',
      'Piper',
      'Hawker',
      'Dassault',
      'Lockheed',
      'Beechcraft',
      'Gulfstream',
      'Cessna',
      'Piper',
      'Hawker',
      'Dassault',
      'Lockheed',
      'Beechcraft',
      'Gulfstream',
    ],
  });

  const t = d.get('time', () => '7');

  return {
    plugins: [],
    processors: [
      {
        name: 'hello',
        files: [
          {
            root: null,
            glob: '**/*.*',
            exclude: [],
            type: GlobType.Template,
          },
        ],
        config: {
          Name: name,
          Age: age,
          Food: food,
          Color: color,
          Beef: beefOk,
          Pork: porkOk,
          Birthday: birthday,
          MumBirthday: mumBirthday,
          Password: password,
          Pin: pin,
          Car: car,
          Plane: plane,
          Time: t,
        },
      },
    ],
  } satisfies Cyan;
});
